wire-ios-sync-engine/Source/Calling/MLSConferenceStaleParticipantsRemover.swift

```diff
@@ -0,0 +1,218 @@
+
+    func cancelPendingRemovals() {
+        timerManager.cancelAllTimers()
+    }
+
+    // MARK: - Participants change handling
+
+    private func process(input: MLSConferenceParticipantsInfo) {
+
+        guard let subconversationMembers = subconversationMembers(for: input.subconversationID) else {
+            return
+        }
+
+        let newAndChangedParticipants = newAndChangedParticipants(
+            between: previousInput?.participants ?? [],
+            and: input.participants
+        )
+
+        newAndChangedParticipants.excludingSelf(in: context).forEach {
+
+            guard let clientID = MLSClientID(callParticipant: $0) else {
+                return
+            }
+
+            switch (subconversationMembers.contains(clientID), $0.state) {
+            case (true, .connecting):
+                remove(
+                    client: clientID,
+                    from: input.subconversationID,
+                    after: removalTimeout
+                )
+            case (false, _), (_, .connected):
+                cancelRemoval(for: clientID)
+            default: break
+            }
+        }
+
+        previousInput = input
+    }
+
+    private func newAndChangedParticipants(between previous: [CallParticipant], and current: [CallParticipant]) -> [CallParticipant] {
+        var newAndChanged = [CallParticipant]()
+
+        let previousStates = Dictionary(uniqueKeysWithValues: previous.map { ($0.userId, $0.state) })
+
+        current.forEach { participant in
+            if let previousState = previousStates[participant.userId], previousState != participant.state {
+                newAndChanged.append(participant)
+            } else if previousStates[participant.userId] == nil {
+                newAndChanged.append(participant)
+            }
+        }
+
+        return newAndChanged
+    }
+
+    // MARK: - Helpers
+
+    private func subconversationMembers(for groupID: MLSGroupID) -> [MLSClientID]? {
+        do {
+            return try mlsService.subconversationMembers(for: groupID)
+        } catch {
+            logger.warn("failed to fetch subconversation members: \(String(describing: error))")
+            return nil
+        }
+    }
+
+    private func remove(
+        client clientID: MLSClientID,
+        from groupID: MLSGroupID,
+        after duration: TimeInterval
+    ) {
+        do {
+            try timerManager.startTimer(
+                for: clientID,
+                duration: duration,
+                completion: { [weak self] in
+                    guard let self = self else { return }
+
+                    self.remove(
+                        client: clientID,
+                        from: groupID
+                    )
+                }
+            )
+
+            logger.info("started timer for removal of stale participant (clientdID: \(clientID), groupID: \(groupID))")
+        } catch {
+            // timer already exists, do nothing
+        }
```

Comment:
Is there a better way to check that we already have a timer?

Reply:
I see that the error thrown here is `timerAlreadyExists`, I suggest we catch it explicitly.

Reply:
done ✅